## RSA - The Project

Name:  Taylor Larrechea

Moodle Email: tala4156@colorado.edu

Name your file "cuidentikey".ipynb  This means YOUR identikey. 

**Do NOT delete instructions or pre-existing comments in the code.** 


<hr />

# Table of Contents

### 1. Introduction  (5 points)

### 2. RSA Code Package    (10 points)
###### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.1 Basic tool set
###### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.2 First tool set
###### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2.3 Second tool set

### 3. RSA More Code  (10 points)

###### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  Encode
###### &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  Decode

### 4. Create a demo of encoding and decoding a message that highlights how your code works and the steps involved.   (10 points)

### 5. Results of exchanging codes with others.  (5 points)

### 6. Narrative (25 points)

### 7. Why FME?  (5 points)

### 8. How to break codes  (5 points)

### 9. Code breaking complete examples  (5 points)

### 10. Custom code feature / Advanced options  (15 points) 








### 1. Introduction to your RSA Package and Project.  Give an overview of your project and what you have learned. 

<span style="color:red;">Introduction</span>

RSA is a data encryption / decryption technique that is widely used in multiple facets in society for the safe transfer of data and information. RSA incorporates many mathematical procedures ranging from fast modular exponentiation (FME) to the extended Euclidean algorithm (EEA) and more. This project aims to generate encrypted messages that can be decrypted with the use of public key exchanges between parties. With access to these keys, one can decrypt encrypted messages in a safe manner where only the intended parties have access to the information. I will walk through the necessary components that were used to build this project, demonstrate its use, and discuss the results from this project. Furthermore, I will discuss possible pitfalls that exist in this project and how they can be improved.

### 2.1
#### Basic tool set

These are functions that you'll need to pre-process the messages before the messages are encoded and decoded by the RSA algorithm. That is the reason we will be defining them first.



<span style="color:red;">Basic Tool Set Code</span>

In [21]:
# Convert_Text - Converts a string to an array of integers that correspond to the ascii code for a given value
# Input:
#   _string - String that is to be converted to an array of ascii integers
# Algorithm:
#   * Create an empty array where the ascii digits are supposed to be appended to
#   * Iterate over the string and append the corresponding ascii value to integer_list
#       * Append the ascii code to integer_list
#   * Return the array
# Output:
#   integer_list - Array of integer values pertaining to the ascii of a letter or character
def Convert_Text(_string):
    """
    Define this function such that it takes in a simple 
    string such as "hello" and outputs the corresponding
    standard list of integers (ascii) for each letter in the word hello.
    For example:
    _string = hello
    integer_list = [104, 101, 108, 108, 111]
    
    You may use "ord()"
    """
    integer_list = []
    for i in _string:
        integer_list.append(ord(i))
    return integer_list

In [22]:
# Convert_Num - Converts an array of numbers representing ascii numbers for a given character
# Input:
#   _list - Array of integers that represent ascii codes for a given character
# Algorithm:
#   * Create an empty string where characters are to be appended to the string
#   * Iterate over the array and append the characters to the string
#       * Append the character in _list to _string
#   * Return the string
# Output:
#   _string - String that represents the converted array to string
def Convert_Num(_list):
    """
    Do the opposite of what you did in the Convert_Text
    function defined above.
    
    Define this function such that it takes in a list of integers
    and outputs the corresponding string (ascii).
    
    For example:
    _list = [104, 101, 108, 108, 111]
    _string = hello
    """
    _string = ''
    for i in _list:
        _string += chr(i)
    return _string

In [23]:
# Convert_Binary_String - Converts a decimal number to binary
# Input:
#   _int - Integer value that is to represent a number in decimal that is to be converted to binary
# Algorithm:
#   * Create an empty string that will represent the binary number
#   * While _int is greater than 0
#       * Calculate the binary digit by taking the modulo of _int and 2
#       * Update the string value with this digit
#       * Update _int by calculating the floor of _int divided by 2
#   * Reverse the order of the string
#   * Return the string
# Output:
#   _string - String value that is equal to the binary number of the original integer _int
def Convert_Binary_String(_int):
    """
    Here, you need to define a function that converts an integer to
    a string of its binary expansion.
    
    For example:
    _int = 345
    bits = 101011001
    """
    _string = ''
    while(_int > 0):
        bindig = _int % 2
        _string += str(bindig)
        _int = _int // 2
    _string = _string[::-1]
    return _string

Now that you're done with the basic toolset we'll move on to the first tool set which is actually involved in the RSA system.

### 2.2 
#### First tool set.



<span style="color:red;">First Tool Set Code</span>

In [24]:
# FME - A function to calculate Fast Modular Exponentiation (b^n mod m)
# Input:
#   b - Integer value that represents the base of of the FME (b) in (b^n mod m)
#   n - Integer value that represents the exponent value in FME (n) in (b^n mod m)
#   m - Integer value that represents the value that is being modulo'd against in FME (m) in (b^n mod m)
# Algorithm:
#   * Calculate the binary representation of n with Convert_Binary_String
#   * Reverse the binary number so that the last digit appears first in the array
#   * Set the return value (x) to 1
#   * Calculate power by taking the modulo of b and m
#   * Traverse the string of binary digits up to the length of the binary number
#       * Convert the string to an integer and store it in binVal
#       * If the current digit is a one, calculate x with (x * power) % m
#       * Update the power variable with (power * power) % m
#   * Return x after the for loop finishes completion
# Output:
#   x - Integer value that is returned from the expression of (b^n mod m)
def FME(b,n,m):
    """
    1. Using the fast modular exponentiation algorithm,
    the below function should return b**n mod m.
    As described on page 254. (however, you may input the exponent and 
    then convert it to a string - the book verison imports the binary expansion)  
    2. You should use the function defined above Convert_Binary_String() if using the book algorithm.
    3. For this block you MUST use one of the 3 methods offered in the Mastery Workbook.
    4. Any method using bit-shifting or copied from the internet (even changing varibale names) will result in a 0.
    5. **If you are completely stuck, you may use pow() with a 10pt penalty.**

    You may use the function you developed in your Mastery Workbook - but be sure it is your own
    work, commented, etc. and change inputs as needed.
    """  
    DecInBin = Convert_Binary_String(n)
    DecInBin = DecInBin[::-1]
    x = 1
    power = b % m
    for i in DecInBin:
        binVal = int(i)
        if (binVal == 1):
            x = (x * power) % m
        power = (power * power) % m
    return x

In [25]:
# Euclidean_Alg - Greatest Common Divisor or Euclidean algorithm for two numbers a and b
# Input:
#   a - Integer value that represents the larger of the two values in the EEA calculation
#   b - Integer value that represents the smaller of the two values in the EEA calculation
# Algorithm:
#   * Assign values to the initial of a and b with m_0 and b_0
#   * Assign values to the initial Bezout coefficients s_1,t_1 with 1 and 0
#   * Assign values to the updated Bezout coefficients s_2,t_2 with 0 and 1
#   * Execute the while loop until b is less than or equal to 0 (usually equal to 0)
#       * Calculate the modulo of a and b and assign it to k
#       * Calculate the integer division of a and b and assign it to q
#       * Update a to the current value of b and b to to the current value of k
#       * Calculate new values for the Bezout coefficients S_1,T_1 with s_2,t_2
#       * Calculate the new updated Bezout coefficients S_2,T_2 with s_1 - q * s_2, t_1 - q * t_2
#       * Update the Bezout coefficients (which will eventually be returned) s_1,t_1 with S_1,T_1
#       * Update the updated Bezout coefficients s_2,t_2 with S_2,T_2
#   * Return the GCD (a) and the Bezout coefficients (s_1,t_1) as an array after the loop
# Output:
#   This algorithm returns an array of values related to the Extended Euclidean Algorithm
#   a - This is the GCD of the two original numbers a and b
#   s_1 - This is the first Bezout coefficient (s) that is returned
#   t_1 - This is the second Bezout coefficient (t) that is returned
def Euclidean_Alg(a,b):
    """
    1. Calculate the Greatest Common Divisor of a and b.
    
    2. This version should have only postive inputs and outputs.
    
    3. The function must return a single integer ('x') which is
    the gcd of a and b.
    
    
    """
    a_0,b_0 = a,b
    s_1,t_1 = 1,0
    s_2,t_2 = 0,1
    while (b > 0):
        k = a % b
        q = a // b
        a = b
        b = k
        S_1,T_1 = s_2,t_2
        S_2,T_2 = s_1 - q * s_2, t_1 - q * t_2
        s_1,t_1 = S_1,T_1
        s_2,t_2 = S_2,T_2
    return [a,s_1,t_1]

### 2.3
#### Second tool set

Here we will implement the core of the RSA cryptosystem. The functions below will generate the public and private key pairs which will then be used to create a ciphertext using the public key and then decode the same using the pirvate key.



<span style="color:red;">Second Tool Set Code</span>

In [26]:
# Find_Public_Key_e - This function finds a public key e such that it is relatively prime to (p-1)(q-1)
# Input:
#   p - First prime number
#   q - Second prime number
# Algorithm:
#   * Calculate both (p-1) and (q-1) and store them as P and Q
#   * Calculate the product of p and q and store it as n
#   * Initialize e to be 2
#   * While the gcd(e,P*Q) is not 1 and e is not equal to both p and q
#       * Increment e by 1
#   * Return e and n in an array
# Output:
#   This function returns the public key e and n
def Find_Public_Key_e(p, q):
    """
    Implement this function such that
    it takes 2 primes p and q.
    
    Use the gcd function that you have 
    defined before.
    
    The function should return 2 elements as follows:
    public key: n
    public key: e
    
    HINT: this function will run a loop to find e such 
    that e is relatively prime to (p - 1) (q - 1) 
    and not equal to p or q.
    
    NOTE: There are a number of ways to implement this key feature. 
    You, as the coder, can choose to how to achieve this goal.
    
    """
    P = p - 1
    Q = q - 1
    n = p * q
    e = 2
    while (Euclidean_Alg(e,P*Q)[0] != 1 and e != p and e != q):
        e += 1
    return [e,n]

In [27]:
# Find_Private_Key_d - This function returns the private key d from a public key e and two prime numbers p and q
# Input:
#   e - Public key that is calculated with Find_Public_Key_e
#   p - First prime number
#   q - Second prime number
# Algorithm:
#   * Calculate (p-1) and (q-1) and store them as P and Q respectively
#   * Calculate the modular inverse (the Bezout coefficient) of e and P*Q and store it as d
#   * If d is negative, add the modulo to it until it is positive
#   * Return d
# Output:
#   This function returns the public key d
def Find_Private_Key_d(e, p, q):
    """
    Implement this function to find the decryption exponent d, 
    such that d is the modular inverse of e. 
    
    This will use the Extended Euclidean Algorithm
    
    This function should return the following:
    d: the decryption component.
    
    This is not a single action, and there are multiple methods to create this. 
    
    You may create a helper function or have all code within this function.
    
    Plan ahead before coding this.

    """
    P = p - 1
    Q = q - 1
    d = Euclidean_Alg(e,P*Q)[1]
    if (d < 0):
        while(d < 0):
            d += P*Q
    return d

### 3.
#### Putting things all together.

1. In this part, you will define two functions `Encode` and `Decode` which will use the public and private keys that you calculated using the above 2 functions in the second toolset.
2. Using the public key, the `Encode` function will encode a message and generate the corresponding cipher_text.
3. Using the private key, the `Decode` function will decode a ciper_text and recover the original message.



<span style="color:red;">Encode And Decode Algorithms</span>

In [28]:
# Encode - Encodes a string of text from a public key e and number n
# Input:
#   n - Integer value that is calculated with Find_Public_Key_e
#   e - Integer value that is calculated with Find_Public_Key_e
#   message - Message that is fed into the algorithm to by encoded
# Algorithm:
#   * Create an empty array called cipher_text
#   * Iterate over the characters in the message
#       * Calculate the integer value of char from Convert_Text
#       * Encode charNum with the use of FME and the value of e and n with charNum
#       * Append this value to cipher_text
#   * Return cipher_text
# Output:
#   This function returns an array of integers that represent encoded values from message
def Encode(n,e,message):
    """
    Here, the message will be a string of characters.
    Use the function Convert_Text from 
    the basic tool set and get a list of numbers.
    
    Encode each of these numbers using n and e and
    return the encoded cipher_text.
    """
    cipher_text = []
    for char in message:
        charNum = Convert_Text(char)[0]
        encoded = FME(charNum,e,n)
        cipher_text.append(encoded)
    return cipher_text

In [29]:
# Decode - Decodes a message from an array of integers with private key d, n, and cipher_text
# Input:
#   n - Integer value that is calculated with Find_Public_Key_e
#   d - Integer value that is calculated with Find_Private_Key_d
#   cipher_text - Array of integers that represents an encoded message
# Algorithm:
#   * Create an empty container for the message that is to be returned
#   * Create an empty array to store the converted integers to strings in
#   * Iterate over the cipher_text array
#       * Calculate the ascii code for the given string with FME
#       * Append this ascii code to empty
#   * Decode the ascii values with Convert_Num and assign the result to message
#   * Return message
# Output:
#   This function returns the decoded message from cipher_text with private key d and n
def Decode(n,d,cipher_text):
    """
    Here, the cipher_text will be a list of integers.
    First, you will decrypt each of those integers using 
    n and d.
    
    Later, you will need to use the function Convert_Num from the 
    basic toolset to recover the original message as a string. 
    
    """
    message = ''
    empty = []
    for num in cipher_text:
        numVal = FME(num,d,n)
        empty.append(numVal)
    message = Convert_Num(empty)
    return message

### 4.  Create a Demo 
This is a **step-by-step** guide to using your code with a specific example that we can follow using the functions you have created above in a mix of code and text blocks. 

**This is essentially a test that demonstates that your code works and how it works. You will call the functions above. You are not using a main function or custom feature here - just calling the functions.**

Imagine you are showing me how you code works in person. 
Your function calls show how it works with an example.
You text s what you would say to me.

It is mix of code blocks and **text blocks** and must discuss all three parts of the project for full credit. You must **show and explain your method and demonstrate (by calling your function as an example.** for generating keys and how your encoding and decoding functions work. 


* Generate keys. 

* Encode a message.

* Decode the same message. 


<span style="color:red;">Demo</span>

In [30]:
testMessage = "This is a test of RSA for this project." # Test message
p = 1599380351 # Prime number 1
q = 3364670173 # Prime number 2
Public_Key = Find_Public_Key_e(p,q) # Generate public keys from p and q
for i in range(len(Public_Key)):
    if (i == 0):
        print(f"Public key e = {Public_Key[i]}")
    else:
        print(f"n value = {Public_Key[i]}")
Private_Key = Find_Private_Key_d(Public_Key[0],p,q) # Generate private keys from e, p, and q
print(f"Private key d = {Private_Key}")
Encoded = Encode(Public_Key[1],Public_Key[0],testMessage) # Generate encoded text
print(f"Encoded text = {Encoded}")
Decoded = Decode(Public_Key[1],Private_Key,Encoded)
print(f"Decoded text = {Decoded}")

Public key e = 7
n value = 5381387362291970723
Private key d = 1537539244950834343
Encoded text = [29509034655744, 131593177923584, 140710042265625, 266001988046875, 34359738368, 140710042265625, 266001988046875, 34359738368, 80798284478113, 34359738368, 282621973446656, 107213535210701, 266001988046875, 282621973446656, 34359738368, 207616015289871, 114868566764928, 34359738368, 24928547056768, 27136050989627, 4902227890625, 34359738368, 114868566764928, 207616015289871, 250226879128704, 34359738368, 282621973446656, 131593177923584, 140710042265625, 266001988046875, 34359738368, 221068140740608, 250226879128704, 207616015289871, 150363025899136, 107213535210701, 93206534790699, 282621973446656, 435817657216]
Decoded text = This is a test of RSA for this project.


<span style="color:red;">Demo Explanation</span>


The above involves the test message "This is a test of RSA for this project.". The prime numbers used in this example are p = 1599380351 and q = 3364670173. We then

* Calculate The Public Key e And n Value:
    * The public key in this case is e = 7
    * The value for n in this case is n = 5381387362291970723
* Calculate The Private Key d:
    * The private key in this case is d = 1537539244950834343
* Encode The Message:
    * The array of integers that represents this encoded message is: [29509034655744, 131593177923584, 140710042265625, 266001988046875, 34359738368, 140710042265625, 266001988046875, 34359738368, 80798284478113, 34359738368, 282621973446656, 107213535210701, 266001988046875, 282621973446656, 34359738368, 207616015289871, 114868566764928, 34359738368, 24928547056768, 27136050989627, 4902227890625, 34359738368, 114868566764928, 207616015289871, 250226879128704, 34359738368, 282621973446656, 131593177923584, 140710042265625, 266001988046875, 34359738368, 221068140740608, 250226879128704, 207616015289871, 150363025899136, 107213535210701, 93206534790699, 282621973446656, 435817657216]
* Decode The Message:
    * The decoded message is then: This is a test of RSA for this project.

Here I have correctly generated public and private keys, encoded a message, and in turn decoded the message from these keys, the encoded message, and the n value.

### 5. Code Exchange
Now that your code is working and you are able read and write message, include 3 complete examples of exhanging code from Piazza here. Put BOTH the code and solutions here. 

<span style="color:red;">Derek Larson Response</span>

The public key and value of n are: n = 1649, e = 115

The private key and value of n are: n = 1649, d = 187

The encoded message is: [155, 203, 1095, 23, 497, 825, 497, 203, 203, 825, 27, 1105, 825, 1095, 1096, 27, 497, 825, 1095, 1096, 27, 716, 188, 825, 1530, 1154, 317, 890, 497, 1507, 825, 1147, 1096, 776, 1095, 23, 497, 825, 722, 1154, 536, 317, 825, 1105, 776, 815, 1154, 317, 27, 1095, 203, 825, 1476, 27, 571, 571, 776, 825, 1095, 1154, 1476, 1476, 27, 716, 188, 504]

The code for generating a response to this can be seen below:

In [31]:
secretMessage = [155, 203, 1095, 23, 497, 825, 497, 203, 203, 825, 27, 1105, 825, 1095, 1096, 27, 497, 825, 1095, 1096, 27, 716, 188, 825, 1530, 1154, 317, 890, 497, 1507, 825, 1147, 1096, 776, 1095, 23, 497, 825, 722, 1154, 536, 317, 825, 1105, 776, 815, 1154, 317, 27, 1095, 203, 825, 1476, 27, 571, 571, 776, 825, 1095, 1154, 1476, 1476, 27, 716, 188, 504]
print(Decode(1649,187,secretMessage))
responseMessage = "Definitely jalepenos. If not jalepenos, banana peppers!"
print(Encode(1649,115,responseMessage))

Let's see if this thing works. What's your favorite pizza topping?
[816, 203, 1105, 27, 716, 27, 1095, 203, 607, 722, 825, 489, 776, 607, 203, 1476, 203, 716, 1154, 497, 1507, 825, 1315, 1105, 825, 716, 1154, 1095, 825, 489, 776, 607, 203, 1476, 203, 716, 1154, 497, 31, 825, 1262, 776, 716, 776, 716, 776, 825, 1476, 203, 1476, 1476, 203, 317, 497, 1308]


<span style="color:red;">Thomas Dunn Response</span>

The public key and value of n are: n = 94043, e = 77893

The private key and value of n are: n = 94043, d = 59245

The encoded message is: [4601, 88511, 36234, 52588, 27051, 75635, 15375, 52588, 68716, 67855, 75635, 27451, 52588, 70258, 51010, 60685, 70258, 52588, 70258, 51010, 20485, 52588, 86547, 60685, 9881, 20485, 52588, 82631, 73433, 47782, 52588, 56805, 74324, 75635, 49971, 74324, 60685, 9881, 52588, 27451, 75635, 74324, 68716, 86547, 52588, 27451, 88511, 70258, 51010, 52588, 36234, 88511, 92225, 92225, 20485, 74324, 20485, 67855, 70258, 52588, 45073, 60685, 67855, 49971, 15375, 60685, 49971, 20485, 86547, 92575, 52588, 52588, 59141, 75635, 15375, 52588, 23650, 15375, 86547, 70258, 52588, 67855, 20485, 20485, 36234, 52588, 45073, 60685, 74324, 49971, 20485, 74324, 52588, 56805, 74324, 88511, 9881, 20485, 86547, 83986, 52588, 52588, 54495, 14902, 33964, 52588, 77050, 12248, 29183, 16531, 51881, 47024, 35193, 51881, 46292, 67886, 35193, 34047, 83986]

The code for generating a response to this can be seen below:

In [32]:
secretMessage = [4601, 88511, 36234, 52588, 27051, 75635, 15375, 52588, 68716, 67855, 75635, 27451, 52588, 70258, 51010, 60685, 70258, 52588, 70258, 51010, 20485, 52588, 86547, 60685, 9881, 20485, 52588, 82631, 73433, 47782, 52588, 56805, 74324, 75635, 49971, 74324, 60685, 9881, 52588, 27451, 75635, 74324, 68716, 86547, 52588, 27451, 88511, 70258, 51010, 52588, 36234, 88511, 92225, 92225, 20485, 74324, 20485, 67855, 70258, 52588, 45073, 60685, 67855, 49971, 15375, 60685, 49971, 20485, 86547, 92575, 52588, 52588, 59141, 75635, 15375, 52588, 23650, 15375, 86547, 70258, 52588, 67855, 20485, 20485, 36234, 52588, 45073, 60685, 74324, 49971, 20485, 74324, 52588, 56805, 74324, 88511, 9881, 20485, 86547, 83986, 52588, 52588, 54495, 14902, 33964, 52588, 77050, 12248, 29183, 16531, 51881, 47024, 35193, 51881, 46292, 67886, 35193, 34047, 83986]
print(Decode(94043,59245,secretMessage))
responseMessage = "I did not know this! I imagine that you have to use different ascii codes for characters in other languages. But this is pretty cool regardless!"
print(Encode(94043,77893,responseMessage))

Did you know that the same RSA program works with different languages?  You just need larger primes!  你好！ Здравствуйте!
[60136, 52588, 36234, 88511, 36234, 52588, 67855, 75635, 70258, 52588, 68716, 67855, 75635, 27451, 52588, 70258, 51010, 88511, 86547, 83986, 52588, 60136, 52588, 88511, 9881, 60685, 49971, 88511, 67855, 20485, 52588, 70258, 51010, 60685, 70258, 52588, 27051, 75635, 15375, 52588, 51010, 60685, 18433, 20485, 52588, 70258, 75635, 52588, 15375, 86547, 20485, 52588, 36234, 88511, 92225, 92225, 20485, 74324, 20485, 67855, 70258, 52588, 60685, 86547, 6433, 88511, 88511, 52588, 6433, 75635, 36234, 20485, 86547, 52588, 92225, 75635, 74324, 52588, 6433, 51010, 60685, 74324, 60685, 6433, 70258, 20485, 74324, 86547, 52588, 88511, 67855, 52588, 75635, 70258, 51010, 20485, 74324, 52588, 45073, 60685, 67855, 49971, 15375, 60685, 49971, 20485, 86547, 61112, 52588, 51122, 15375, 70258, 52588, 70258, 51010, 88511, 86547, 52588, 88511, 86547, 52588, 56805, 74324, 20485, 70258, 70258, 27

<span style="color:red;">Personal Post</span>

The original post is: I'm a team from Mile High, My uniforms are orange and blue, My fans often chant In-Com-Plete, A franchise orchestrated with a coach that is new. With a horse in my name, I proudly stand, In Denver's heart, across the land.

The primes that were used to generate these keys and the message are:

p = 26107057872784305673

q = 75432624068748034463

The keys that are generated for this are:

n = 1969323882058787277298186868172930408599, e = 5, d = 393864776411757455439329437246279613693

The encoded message is then: [2073071593, 90224199, 15386239549, 33554432, 8587340257, 33554432, 21003416576, 10510100501, 8587340257, 15386239549, 33554432, 11040808032, 19254145824, 16850581551, 15386239549, 33554432, 2706784157, 12762815625, 14693280768, 10510100501, 33554432, 1934917632, 12762815625, 11592740743, 12166529024, 164916224, 33554432, 4984209207, 12762815625, 21003416576, 12166529024, 33554432, 8587340257, 33554432, 12166529024, 16850581551, 19254145824, 20113571875, 10510100501, 33554432, 12762815625, 16105100000, 33554432, 15386239549, 25937424601, 33554432, 16105100000, 8587340257, 15386239549, 10510100501, 164916224, 33554432, 2073071593, 33554432, 17623416832, 19254145824, 16850581551, 21924480357, 10000000000, 14693280768, 25937424601, 33554432, 20113571875, 21003416576, 8587340257, 16105100000, 10000000000, 164916224, 33554432, 2073071593, 16105100000, 33554432, 1453933568, 10510100501, 16105100000, 22877577568, 10510100501, 19254145824, 90224199, 20113571875, 33554432, 12166529024, 10510100501, 8587340257, 19254145824, 21003416576, 164916224, 33554432, 8587340257, 9509900499, 19254145824, 16850581551, 20113571875, 20113571875, 33554432, 21003416576, 12166529024, 10510100501, 33554432, 14693280768, 8587340257, 16105100000, 10000000000, 205962976]

In [33]:
secretMessage = "I'm a team from Mile High, With a horse in my name, I proudly stand, In Denver's heart, across the land."
p = 26107057872784305673
q = 75432624068748034463
publicKey = Find_Public_Key_e(p,q)
e = publicKey[0]
n = publicKey[1]
print(f"Public key e = {e}, n value = {n}")
d = Find_Private_Key_d(e,p,q)
print(f"Private key d = {d}")
encoded = Encode(n,e,secretMessage)
print(f"Encoded message: {encoded}")
print(f"Decoded message: {Decode(n,d,encoded)}")

Public key e = 5, n value = 1969323882058787277298186868172930408599
Private key d = 393864776411757455439329437246279613693
Encoded message: [2073071593, 90224199, 15386239549, 33554432, 8587340257, 33554432, 21003416576, 10510100501, 8587340257, 15386239549, 33554432, 11040808032, 19254145824, 16850581551, 15386239549, 33554432, 2706784157, 12762815625, 14693280768, 10510100501, 33554432, 1934917632, 12762815625, 11592740743, 12166529024, 164916224, 33554432, 4984209207, 12762815625, 21003416576, 12166529024, 33554432, 8587340257, 33554432, 12166529024, 16850581551, 19254145824, 20113571875, 10510100501, 33554432, 12762815625, 16105100000, 33554432, 15386239549, 25937424601, 33554432, 16105100000, 8587340257, 15386239549, 10510100501, 164916224, 33554432, 2073071593, 33554432, 17623416832, 19254145824, 16850581551, 21924480357, 10000000000, 14693280768, 25937424601, 33554432, 20113571875, 21003416576, 8587340257, 16105100000, 10000000000, 164916224, 33554432, 2073071593, 16105100000,

### 6. Write a Narrative (TWO pages of meaningful narrative about your process - not including the examples - about 800 words) here in this block.

* Describe your process. Did you plan ahead? Did you wing it? Do you wish you'd planned ahead more?

* Describe the results of exchanging keys and codes with classmates. Did it work the first time or did you need to make adjustments.Describe your process. Did you plan ahead? Did you wing it? Do you wish you'd planned ahead more?

* What was most challenging? 

* Who helped you with your project?  Which resources were most helpful?

* What was your Best Mistake? (funniest? most frustrating? the one you learned the most from?)




<span style="color:red;">RSA Narrative</span>

During the process of completing this project I had to formulate a game plan as to how to complete each part of the project. I first started making the necessary components of the project by completing the problems in the mastery workbooks. I originally made an FME algorithm that was recursive (of which I did not initially know was acceptable) and then learned that it was not allowed. So, I had to go back to the drawing board and rewrite my algorithm so that it matched the one from Rosen. I am more inclined to use recursive algorithms when possible because of the difficulty for outsiders who may plan to copy my code at some point. Plus I really think recursion is cool and thus why I prefer.

After completing the FME algorithm I moved on to completing the EEA algorithm. Both of these algorithms are imperative to RSA's success and so I spent a lot of time making sure that they did what they needed to do. Upon completing each algorithm, I first checked simple problems by hand to make sure that they were working in the way that they were intended to work. This enabled me to make sure that my algorithms were correct. I then scoured the internet to find resources that would allow me to check more complicated examples. For the FME algorithm I used a resource at (https://www.omnicalculator.com/math/power-modulo) to work on more complicated examples. Once the results that I input online were matching with that of my algorithm I had full confidence that my algorithm was correct.

I then moved on to checking that my EEA algorithm was working the way that I had hoped. At first I worked with simple examples and checked again that what I was putting into my EEA was correct by checking it by hand. Once I got the simple examples were working (I primarily checked my examples with my work from the mastery workbook problems) and then moved on to finding resources online that could handle more complicated computations. For the EEA I used a resource at (https://planetcalc.com/8586/) to check more complicated examples with my custom EEA. After completing this process I moved on to working on the functions inside the Jupyter notebook so that they could all be pieced together. 

After the FME and EEA algorithms were completed I moved on to the algorithms that were inside the mastery workbook. Before I put my finished code for each algorithm I made a separate Python file that contained the necessary components from previous algorithms so that they could be tested with the new one that I was currently working on. The first three algorithms, Convert_Text, Convert_Num, and Convert_Binary_String were rather easy for me to complete. At first I was trying to code Convert_Text manually instead of using ord() and this was a big mistake for me. I was trying to construct arrays, dictionaries, and other containers that could be used to store ascii codes with certain containers and that was a nightmare for me. I then went back to the comments in the notebook and researched what ord did in the context in Python and I was able to complete the first algorithm after that. Completing Convert_Num was rather simple with the use of chr and then the last basic algorithm that I needed to complete was Convert_Binary_String and this leads to my funniest mistake that I made in the project.

I am good at a lot of things but one thing that I am subpar in is reading directions. This proved to be a big achilles heal to me as I spent a copious amount of time hand coding a decimal to binary algorithm. I cannot understate how long I spent on this one algorithm alone. I had no idea how to get it to work, I had no idea what to store the result in at first, and I had no idea how to even begin. So I referenced Rosen and was able to get on track with the pseudo code that was provided to me. Upon completing the algorithm originally I noticed that my digits were in reverse order. Thankfully I had spent a lot of time in Python in my past when I was a physicist and was able to quickly reverse the string that I was working with inside the algorithm. After completing the algorithm and then putting it inside my notebook I noticed that we were allowed to use bin in this function and I let out an audible groan after all the time that I had spent coding my algorithm. After this algorithm I made sure to read the instructions before starting my algorithms and the rest of the product went pretty smoothly from there.

I then proceed to complete the remaining algorithms with a lot less effort and was able to get my code working where I could accurately produce keys, encode a message, and decode a message that was given to me. At this point I was now ready to start exchanging keys and codes with classmates. Upon getting all of the algorithms complete the first message that I attempted to decrypt was the one that Professor Stade posted that was supposed to say "Hello Earth". Upon putting all of the necessary information into my IDE I was jumping with joy to see that my code was actually working. This then enabled me to go to other threads and check that my code was working with other people's posts. In the first thread that involved students posting both their public and private keys with their cipher text, I was able to decode the messages with relative ease and it wasn't too hard for me to produce a response. After I had confidence that I was able to decode people's posts with ease I moved on to producing my own post. I wanted to be a little bit different and test the limitations of RSA so I found a resource at (https://bigprimes.org/) that allowed me to produce very large primes. I believe that some people may have been a little apprehensive in decoding my message because of the size of keys but people who did not respond to the post on Piazza said that they were able to decode my post and that gave me confidence that my algorithms were working.

After posting and responding to messages in Piazza in the public thread I moved on to breaking codes in its respective thread. This was more challenging for me because I was unable to decrypt some peoples posts but upon with consulting with other students I found out that they were having the same problem. I had to baseline my code breaking with an example that I knew was most likely correct. So, I broke the code that was posted by Professor Stade and I was able to successfully break it so I knew what I was doing was most likely correct. I did have a mistake in my Find_Private_Key_d function in how I was handling the generation of d when the original was negative. Originally I was just returning the other Bezout coefficient but upon consulting with Professor Stade I realized that was incorrect and I had to go back to the drawing board. I then changed my code so that it added the modulo value to d until it was positive and returned that key after. Upon fixing my code I was able to break other students code and had an efficient and correct process for breaking code.

Overall I really enjoyed this project and was very happy with the results that were produced upon completing it. This is by far one of my favorite projects that I have worked on and I am very happy that I got an opportunity to do so.

### 7. FME Exploration - (CODE INTERVIEW type question)



* Explain in detail why RSA needs to use an FME function and cannot just simply use the Python Mod function % .

* Provide some specific examples and/or include chart and graphs. 

* You may use your arguments/work from the Number Theory Mastery Workbook.




<span style="color:red;">FME Exploration</span>

For starters, the reason why RSA needs to use FME over just the Python Mod function is because RSA works with very large numbers. FME is more efficient than Python's mod function because calculating the modulo of very large numbers requires a lot of computations. When dealing with very large numbers Python's regular mod function would take a long time to compute because of its limitations. On the other hand FME can calculate these results much faster and more securely.

FME also calculates a number raised to another power and then modulo'd with another. This also adds a level of security when producing keys that are hard for others to crack and break the encryption. Because of the previous two reasons; less computations and safer key generation RSA is a more optimal choice for calculating keys and whatnot.

Take for instance the calculation of $15^{1234} \text{ mod } 37 $ with the mod function in Python and FME. 

In [34]:
import time
start_time1 = time.time()
15**1234 % 37
end_time1 = time.time()
elapsed_time1 = end_time1 - start_time1
FME(15,1234,37)
start_time2 = time.time()
end_time2 = time.time()
elapsed_time2 = end_time2 - start_time2

print(f"Regular mod calculation: {elapsed_time1} seconds.")
print(f"FME calculation: {elapsed_time2} seconds.")

Regular mod calculation: 4.1484832763671875e-05 seconds.
FME calculation: 1.0967254638671875e-05 seconds.


We can see for this simple example that FME is around 10 times faster than the regular Python mod function. This discrepancy will increase for larger numbers and that is a big reason why we should use FME instead of just mod in Python.

## CODE BREAKING ##

This section of the project is only available to students exchanging codes by the Friday (midnight) before the project is due.

After you have tested your RSA package yourself, and tested it with classmates by publishing both the private and public keys on Piazza, post 2 messages with just the public keys for people to break on the "Just Public Keys" thread - use both very small n ‘s (under 1000) for practice, and some with a challenge.


**Implementing a factoring algorithm:**

Begin by coding the basic brute force factorization algorithm given in pseudocode below.

Brute Force Factoring
<pre><code>def factorize(n):
    # n is a number, return the smallest factor of n
    for i from 2 to n-1:
        if i divides n:
            return i
        return FALSE
        </code></pre>


<span style="color:red;">Brute Force Factorization Algorithm</span>

In [35]:
# factorize - Factors an integer n
# Input:
#   n - Integer value that is being factored
# Algorithm:
#   * Iterate from 2 to one less than the value of n
#       * Check if the modulo of n and i is zero, if it is return the factors of both
#   * Otherwise continue to the next number
# Output:
#   This function returns an array of integers that represents the factors of n if n isn't prime
def factorize(n):
    for i in range(2,n-1):
        if (n % i == 0):
            return [i,n // i]
        else:
            continue

### 8. How to Break Code. 
* Explain in detail how breaking works in this text block. 
* What are the complete steps? What tools do you need?
* How secure is RSA? Why is it difficult to break codes in RSA? 
* Is our implementation of RSA secure? What are it's flaws?









<span style="color:red;">Breaking Code</span>

To break the code, we need to calculate the factors of n that are generated in the RSA process. If we are able to find the values of p and q (the factors of n), then we can calculate the private key d and thus decrypt a message.

First we start by factoring n. Once we have the factors of n, we call the Find_Private_Key_d function to calculate the private key for this message. When we have d, we proceed to decode the message with the Decode function with the cipher text that was "intercepted" in this process. After this, we know what the message is and have successfully broke the encryption.

RSA is overall very secure compared to other encryption methods. It is hard to break codes in RSA because of the size of n that is typically produced. Even though we could theoretically break n if we had infinite time, that is not practical in the real world. Because of the size of the numbers that we are working with RSA, it would take a very long time to actually break the code. 

I think for the most part that our implementation of RSA is secure. As long as we are using primes p and q that are large enough, the brute force breaking of the code would be very hard. The way that we are generating the cipher text could be considered a flaw. This is because each element in the array of the cipher text is one character. If we made it so that it was more characters (thus being represented by different integers) it could possibly be more difficult to break. This to me is the biggest flaw in our implementation and in turn would be something that I would change to use somewhere else.

### 9. Decode!  Use a mix of text and code blocks

* Decode one example from Piazza in detial showing all the steps.
* Decode and Respond to 3 "just codes with public keys" in Piazza with different sizes of n. Include complete results here.
* What did you notice about larger n s?


<span style="color:red;">Decoding Responses</span>

For large n's, it takes a lot longer to factor them. This feeds into the idea of why RSA is so secure because these values are typically very large and hard to crack.

<span style="color:red;">Decoding Example With All Steps</span>

The first example that I am going to decode is from <span style="color:red;">Ashely Montgomery</span>.

n value = 779

Public key e = 253

Cipher = [752, 643, 32, 254, 643, 345, 32, 100, 643, 114, 373, 32, 523, 604, 32, 211, 461, 461, 523, 439, 523, 643, 604, 32, 439, 643, 32, 439, 211, 373, 523, 604, 46, 32, 555, 773, 211, 20, 20, 671, 20, 272, 32, 73, 159, 32, 20, 643, 120, 32, 100, 313, 211, 439, 32, 461, 643, 32, 254, 643, 345, 32, 461, 643, 272]

I will now show the steps to break this code:

In [36]:
n = 779
e = 253
cipher = [752, 643, 32, 254, 643, 345, 32, 100, 643, 114, 373, 32, 523, 604, 32, 211, 461, 461, 523, 439, 523, 643, 604, 32, 439, 643, 32, 439, 211, 373, 523, 604, 46, 32, 555, 773, 211, 20, 20, 671, 20, 272, 32, 73, 159, 32, 20, 643, 120, 32, 100, 313, 211, 439, 32, 461, 643, 32, 254, 643, 345, 32, 461, 643, 272]
factored = factorize(n)
p = factored[0]
q = factored[1]
print(f"The first factor is {p} and the second factor is {q}.")
d = Find_Private_Key_d(e,p,q)
print(f"The private key for this is d = {d}.")
decoded = Decode(n,d,cipher)
print(f"The decoded message is then: {decoded}")
responseMessage = "Yes I do work! I work at Intel and I manage a tool set that is used to inspect chips that are being manufactured."
encodedResponse = Encode(n,e,responseMessage)
print(f"The encoded response is then: {encodedResponse}")

The first factor is 19 and the second factor is 41.
The private key for this is d = 37.
The decoded message is then: Do you work in addition to taking classes? If so, what do you do?
The encoded response is then: [545, 671, 20, 32, 73, 32, 461, 643, 32, 100, 643, 114, 373, 717, 32, 73, 32, 100, 643, 114, 373, 32, 211, 439, 32, 73, 604, 439, 671, 773, 32, 211, 604, 461, 32, 73, 32, 14, 211, 604, 211, 46, 671, 32, 211, 32, 439, 643, 643, 773, 32, 20, 671, 439, 32, 439, 313, 211, 439, 32, 523, 20, 32, 345, 20, 671, 461, 32, 439, 643, 32, 523, 604, 20, 321, 671, 555, 439, 32, 555, 313, 523, 321, 20, 32, 439, 313, 211, 439, 32, 211, 114, 671, 32, 611, 671, 523, 604, 46, 32, 14, 211, 604, 345, 159, 211, 555, 439, 345, 114, 671, 461, 654]


The private key for this value of n came out to be 37 where the factors of n were found to be 19 and 41.

The decoded message that Ashley posted was "Do you work in addition to taking classes? If so, what do you do?". I then responded with "Yes I do work! I work at Intel and I manage a tool set that is used to inspect chips that are being manufactured.". This then produced the cipher text "[545, 671, 20, 32, 73, 32, 461, 643, 32, 100, 643, 114, 373, 717, 32, 73, 32, 100, 643, 114, 373, 32, 211, 439, 32, 73, 604, 439, 671, 773, 32, 211, 604, 461, 32, 73, 32, 14, 211, 604, 211, 46, 671, 32, 211, 32, 439, 643, 643, 773, 32, 20, 671, 439, 32, 439, 313, 211, 439, 32, 523, 20, 32, 345, 20, 671, 461, 32, 439, 643, 32, 523, 604, 20, 321, 671, 555, 439, 32, 555, 313, 523, 321, 20, 32, 439, 313, 211, 439, 32, 211, 114, 671, 32, 611, 671, 523, 604, 46, 32, 14, 211, 604, 345, 159, 211, 555, 439, 345, 114, 671, 461, 654]".

The second example that I am going to decode is from <span style="color:red;">Professor Stade</span>.

n value = 703

Public key e = 113

Cipher = [544, 472, 233, 242, 250, 555, 591, 550, 376, 364, 210, 165, 526, 47, 555, 95, 476, 210, 242, 165, 95, 233, 526, 210, 376, 95, 233, 242, 364, 210, 242, 224, 376, 95, 364, 233, 47, 242, 210, 555, 376, 165, 472, 242, 555, 410, 242, 250, 95, 233, 233, 47, 145]

In [37]:
n = 703
e = 113
cipher = [544, 472, 233, 242, 250, 555, 591, 550, 376, 364, 210, 165, 526, 47, 555, 95, 476, 210, 242, 165, 95, 233, 526, 210, 376, 95, 233, 242, 364, 210, 242, 224, 376, 95, 364, 233, 47, 242, 210, 555, 376, 165, 472, 242, 555, 410, 242, 250, 95, 233, 233, 47, 145]
factored = factorize(n)
p = factored[0]
q = factored[1]
d = Find_Private_Key_d(e,p,q)
decoded = Decode(n,d,cipher)
print(f"{decoded}")
encodedResponse = "Has anyone looked north of Creed?"
encoded = Encode(n,e,encodedResponse)
print(f"{encoded}")

The Conquistador's treasure is buried south of Creed.
[116, 526, 210, 242, 526, 591, 581, 555, 591, 233, 242, 90, 555, 555, 160, 233, 47, 242, 591, 555, 95, 165, 472, 242, 555, 410, 242, 250, 95, 233, 233, 47, 195]


The third example that I am going to decode is from <span style="color:red;">Thomas Dunn</span>.

n value = 29917061

Public key e = 8684999

Cipher = [22287915, 9078012, 17453490, 17453490, 14264532, 24523837, 985919, 6341542, 17441500, 5831276, 9066565, 12008286, 9078012, 6341542, 9066565, 985919, 6341542, 17441500, 5831276, 29026233, 29334782, 28836202, 29334782, 25597531, 14264532, 6341542, 12008286, 14264532, 6341542, 11639104, 28836202, 7800891, 29026233, 5831276, 9066565, 20457492, 9078012, 15992531, 6341542, 6341542, 24846263, 12106619, 29650399, 23100921, 13584848, 9668251, 18965502, 15992531, 6341542, 6341542, 15697184, 3102918, 28836202, 17453490, 28615097, 985919, 6341542, 15697184, 29026233, 12008286, 12008286, 14264532, 6341542, 15697184, 15040546, 20457492, 29026233, 17453490, 9078012, 22072925]

In [38]:
n = 29917061
e = 8684999
cipher = [22287915, 9078012, 17453490, 17453490, 14264532, 24523837, 985919, 6341542, 17441500, 5831276, 9066565, 12008286, 9078012, 6341542, 9066565, 985919, 6341542, 17441500, 5831276, 29026233, 29334782, 28836202, 29334782, 25597531, 14264532, 6341542, 12008286, 14264532, 6341542, 11639104, 28836202, 7800891, 29026233, 5831276, 9066565, 20457492, 9078012, 15992531, 6341542, 6341542, 24846263, 12106619, 29650399, 23100921, 13584848, 9668251, 18965502, 15992531, 6341542, 6341542, 15697184, 3102918, 28836202, 17453490, 28615097, 985919, 6341542, 15697184, 29026233, 12008286, 12008286, 14264532, 6341542, 15697184, 15040546, 20457492, 29026233, 17453490, 9078012, 22072925]
factored = factorize(n)
p = factored[0]
q = factored[1]
d = Find_Private_Key_d(e,p,q)
decoded = Decode(n,d,cipher)
print(f"{decoded}")
encodedResponse = "What about Johnny Bravo's prime? 13"
encoded = Encode(n,e,encodedResponse)
print(f"{encoded}")

Jenny's prime is probably my favorite.  8675309.  Thanks Tommy Tutone!
[11338336, 3102918, 28836202, 20457492, 6341542, 28836202, 29334782, 29026233, 15040546, 20457492, 6341542, 22287915, 29026233, 3102918, 17453490, 17453490, 14264532, 6341542, 22810953, 5831276, 28836202, 7800891, 29026233, 24523837, 985919, 6341542, 17441500, 5831276, 9066565, 12008286, 9078012, 11094536, 6341542, 25926241, 13584848]


<span style="color:red;">My</span> example that I posted can be seen below.

Original message: Those who say that pineapple doesn't belong on pizza are just wrong!

Prime numbers:

p = 13

q = 17

n value = 689

Public key e = 5

Cipher = [67, 117, 76, 98, 186, 2, 136, 117, 76, 2, 98, 54, 49, 2, 12, 117, 54, 12, 2, 125, 209, 145, 186, 54, 125, 125, 75, 186, 2, 172, 76, 186, 98, 145, 65, 12, 2, 115, 186, 75, 76, 145, 103, 2, 76, 145, 2, 125, 209, 5, 5, 54, 2, 54, 173, 186, 2, 123, 104, 98, 12, 2, 136, 173, 76, 145, 103, 50]

In [39]:
secretMessage = "Those who say that pineapple doesn't belong on pizza are just wrong!"
p = 13
q = 17
publicKey = Find_Public_Key_e(p,q)
e = publicKey[0]
n = publicKey[1]
print(f"n value = {n}, Public key = {e}")
cipher = Encode(n,e,secretMessage)
print(f"Cipher text: {cipher}")
factored = factorize(n)
pp = factored[0]
qq = factored[1]
d = Find_Private_Key_d(e,pp,qq)
decoded = Decode(n,d,cipher)
print(f"Test of decoding: {decoded}")

n value = 221, Public key = 5
Cipher text: [67, 117, 76, 98, 186, 2, 136, 117, 76, 2, 98, 54, 49, 2, 12, 117, 54, 12, 2, 125, 209, 145, 186, 54, 125, 125, 75, 186, 2, 172, 76, 186, 98, 145, 65, 12, 2, 115, 186, 75, 76, 145, 103, 2, 76, 145, 2, 125, 209, 5, 5, 54, 2, 54, 173, 186, 2, 123, 104, 98, 12, 2, 136, 173, 76, 145, 103, 50]
Test of decoding: Those who say that pineapple doesn't belong on pizza are just wrong!



### 10. Custom CODE Feature ##
The custom feature should be a "Stand Alone" feature at the end of the notebook. If you want to use code from above, please copy the functions again as part of your feature so it can be tested independently. Be sure to provide a demonstration showing that it works. See RSA Guide for more information and suggestions. 

**Do not forget the self-grading element.**


<span style="color:red;">Custom CODE Feature</span>

For my custom feature I decided to make a GUI with Python demonstrating encrypting and decrypting messages. I have a video explaining how this works which can be seen at the following link [Fall 2023 RSA Project](https://youtu.be/oU71NMc6mPc). Please make sure to view this video in the highest resolution possible, I recorded this on my large 40" monitor so it can be very pixelated if you view it in a smaller definition.

For this project I primarily used a book to generate a GUI in Python (Create GUI Applications with Python & Qt5 (5th Edition) - Martin Fitzpatrick). Along with this book, I used some other resources for some of the functions / data manipulation within the code.

- [Python Prime Number Generator](https://www.w3resource.com/python-exercises/generators-yield/python-generators-yield-exercise-3.php)
- [Python Convert String To Array](https://stackoverflow.com/questions/15197673/using-pythons-eval-vs-ast-literal-eval)
- [Prime Number Checker](http://www.math.com/students/calculators/source/prime-number.htm)

My video covers the demonstration of this program and I have commented out my code so that it doesn't crash csel when my notebook is ran online. I am including some screen shots of the program that I created as well. These can be seen below.

<p align="center">
  <img src="Images/Image 1.png" alt="Open Screen"/>
</p>

<p align="center">
  <img src="Images/Image 4.png" alt="Encoding" />
</p>

<p align="center">
  <img src="Images/Image 2.png" alt="Decoding" />
</p>

<p align="center">
  <img src="Images/Image 3.png" alt="Encoding Response" />
</p>

It should be noted that when someone is encoding a response where they know the public key and the value of n, this program does not produce a value of the private key and thus why it is equal to 0.

<span style="color:red;">RSA GUI Code</span>

In [40]:
# # Python modules
# import sys
# import random
# import ast
# # RSA modules
# import BasicToolSet as BTS
# import FirstToolSet as FTS
# import SecondToolSet as STS
# import EncryptingDecrpyting as ED
# # PyQT modules
# from PyQt6.QtCore import Qt
# from PyQt6.QtWidgets import (
#     QApplication,
#     QMainWindow,
#     QWidget,
#     QHBoxLayout,
#     QGridLayout,
#     QVBoxLayout,
#     QLabel,
#     QLineEdit,
#     QPushButton,
#     QTextEdit,
# )

# # IsPrime - Checks if a number is prime
# # Input:
# #   n - Integer that is being checked if it is a prime number
# # Algorithm:
# #   * Index from 2 up to the square root of n + 1
# #       * If the modulo of n and i is zero, then the number is not prime
# #       * Return False
# #   * Return True if False is not returned
# # Output:
# #   This function returns a boolean value depending on whether or not an integer is prime
# def IsPrime(n):
#     for i in range(2,int(n**0.5) + 1):
#         if (n % i == 0):
#             return False
#     return True

# # GeneratePrime - Generates a prime number between two values a and b
# # Input:
# #   a - Integer value that represents the lower bound of the two integers
# #   b - Integer value that represents the upper bound of the two integers
# # Algorithm:
# #   * Generate a random number in the range of a and b and assign it to start
# #   * While the number that is generated is not prime
# #       * Increment start by one
# #       * If start reaches the upper bound equal to b
# #           * Double b
# #           * Set start back to a and continue the loop
# #   * Return the prime number start after the loop ends execution
# # Output:
# #   start - Random prime number that is generated between a and b
# def GeneratePrime(a,b):
#     start = random.randrange(a,b)
#     while (IsPrime(start) == False):
#         start += 1
#         if (start > b):
#             b *= 2
#             start = a
#     return start

# # MainWindow - Class that incorporates all of the PyQt components that are necessary for a GUI in this program
# # Constructor
# #   * Set the title of the window
# #   * Set the minimum height and width of the window
# #   * Generate the main layouts of the window
# #   * Generate the nested layouts for the main layouts
# #   * Generate the widgets for the encode column
# #       * Set the default values of each widget
# #       * Set alignments and object names of each widget
# #       * Add widgets to nestedEncodeLayout
# #   * Generate the widgets for the decode column
# #       * Set the default values of each widget
# #       * Set alignments and object names of each widget
# #       * Add widgets to nestedDecodeLayout
# #   * Generate the widgets for the results row
# #       * Set the default values of each widget
# #       * Set alignments and object names of each widget
# #       * Add widgets to mainResultsLayout
# #   * Add the Encode and Decode layouts to mainInputLayout
# #   * Add the Input and Results layouts to mainLayout
# #   * Set the main layout
# #   * Set the central layout
# # ClearInputs - See inline comments
# # DecodeIt - See inline comments
# # EncodeIt - See inline comments
# class MainWindow(QMainWindow):
#     # Constructor
#     def __init__(self):
#         super().__init__()
#         # Title of Window
#         self.setWindowTitle("RSA")
#         # Height and Width of Window
#         self.resize(850,500)
#         self.setMinimumWidth(850)
#         self.setMinimumHeight(400)
#         # Main Layouts
#         mainLayout = QVBoxLayout()
#         mainInputLayout = QHBoxLayout()
#         mainEncodeLayout = QVBoxLayout()
#         mainDecodeLayout = QVBoxLayout()
#         mainResultsLayout = QVBoxLayout()
#         # Nested Layouts
#         nestedEncodeLayout = QGridLayout()
#         nestedDecodeLayout = QGridLayout()
#         # Encode Column Widgets
#         encodeColumnTitle = QLabel("Encode Message")
#         encodeColumnTitle.setAlignment(Qt.AlignmentFlag.AlignHCenter)
#         encodeMessage = QTextEdit()
#         encodeMessage.setPlaceholderText("Enter Message Here...")
#         encodeMessage.setObjectName("encodeMessage")
#         encodePublicKey = QLineEdit()
#         encodePublicKey.setPlaceholderText("Enter Public Key Here...")
#         encodePublicKey.setMaxLength(10000)
#         encodePublicKey.setMinimumWidth(400)
#         encodePublicKey.setObjectName("encodePublicKey")
#         encodenValue = QLineEdit()
#         encodenValue.setPlaceholderText("Enter n Value Here...")
#         encodenValue.setMaxLength(10000)
#         encodenValue.setMinimumWidth(400)
#         encodenValue.setObjectName("encodenValue")
#         encodeButton = QPushButton("Encode Message")
#         encodeButton.setObjectName("encodeButton")
#         encodeButton.clicked.connect(self.EncodeIt)
#         # Add Widgets To nestedEncodeLayout Layout
#         nestedEncodeLayout.addWidget(encodeColumnTitle, 0,0)
#         nestedEncodeLayout.addWidget(encodeMessage,1,0)
#         nestedEncodeLayout.addWidget(encodePublicKey,2,0)
#         nestedEncodeLayout.addWidget(encodenValue,3,0)
#         nestedEncodeLayout.addWidget(encodeButton,4,0)
#         mainEncodeLayout.addLayout(nestedEncodeLayout)
#         # Decode Column Widgets
#         decodeColumnTitle = QLabel("Decode Message")
#         decodeColumnTitle.setAlignment(Qt.AlignmentFlag.AlignHCenter)
#         decodeMessage = QTextEdit()
#         decodeMessage.setPlaceholderText("Enter Message Here...")
#         decodeMessage.setObjectName("decodeMessage")
#         decodePrivateKey = QLineEdit()
#         decodePrivateKey.setPlaceholderText("Enter Private Key Here...")
#         decodePrivateKey.setMaxLength(10000)
#         decodePrivateKey.setMinimumWidth(400)
#         decodePrivateKey.setObjectName("decodePrivateKey")
#         decodenValue = QLineEdit()
#         decodenValue.setPlaceholderText("Enter n Value Here...")
#         decodenValue.setMaxLength(10000)
#         decodenValue.setMinimumWidth(400)
#         decodenValue.setObjectName("decodenValue")
#         decodeButton = QPushButton("Decode Message")
#         decodeButton.setObjectName("decodeButton")
#         decodeButton.clicked.connect(self.DecodeIt)
#         # Add Widgets To nestedDecodeLayout Layout
#         nestedDecodeLayout.addWidget(decodeColumnTitle,0,0)
#         nestedDecodeLayout.addWidget(decodeMessage,1,0)
#         nestedDecodeLayout.addWidget(decodePrivateKey,2,0)
#         nestedDecodeLayout.addWidget(decodenValue,3,0)
#         nestedDecodeLayout.addWidget(decodeButton,4,0)
#         mainDecodeLayout.addLayout(nestedDecodeLayout)
#         # Results Row Widgets
#         resultsRowTitle = QLabel("Encoding / Decoding Results")
#         resultsRowTitle.setAlignment(Qt.AlignmentFlag.AlignVCenter | Qt.AlignmentFlag.AlignHCenter)
#         resultsRowTitle.setObjectName("resultsRowTitle")
#         encodingDecodingResults1 = QTextEdit()
#         encodingDecodingResults1.setPlaceholderText("Encoding / Decoding Results Here...")
#         encodingDecodingResults1.setMaximumHeight(400)
#         encodingDecodingResults1.setObjectName("encodingDecodingResults1")
#         encodingDecodingResults2 = QTextEdit()
#         encodingDecodingResults2.setPlaceholderText("Private Key / Public Key Here...")
#         encodingDecodingResults2.setMaximumHeight(400)
#         encodingDecodingResults2.setObjectName("encodingDecodingResults2")
#         resultsRowClearButton = QPushButton("Clear Entries")
#         resultsRowClearButton.setObjectName("clearResultsButton")
#         resultsRowClearButton.clicked.connect(self.ClearInputs)
#         # Add Widgets to Results Layout
#         mainResultsLayout.addWidget(resultsRowTitle)
#         mainResultsLayout.addWidget(encodingDecodingResults1)
#         mainResultsLayout.addWidget(encodingDecodingResults2)
#         mainResultsLayout.addWidget(resultsRowClearButton)
#         # Add Layouts to mainInputLayout
#         mainInputLayout.addLayout(mainEncodeLayout)
#         mainInputLayout.addLayout(mainDecodeLayout)
#         # Add Layouts to Main Layout
#         mainLayout.addLayout(mainInputLayout)
#         mainLayout.addLayout(mainResultsLayout)
#         # Set Main Layout
#         widget = QWidget()
#         widget.setLayout(mainLayout)
#         # Set The Central Widget
#         self.setCentralWidget(widget)

#     # ClearInputs - Clears the inputs in the window
#     def ClearInputs(self):
#         # Grab children from main window
#         decodeButton = self.findChild(QPushButton,"decodeButton")
#         decodeMessage = self.findChild(QTextEdit,"decodeMessage")
#         decodenValue = self.findChild(QLineEdit,"decodenValue")
#         decodePrivateKey = self.findChild(QLineEdit,"decodePrivateKey")
#         encodeButton = self.findChild(QPushButton,"encodeButton")
#         encodeMessage = self.findChild(QTextEdit,"encodeMessage")
#         encodenValue = self.findChild(QLineEdit,"encodenValue")
#         encodePublicKey = self.findChild(QLineEdit,"encodePublicKey")
#         resultsTitle = self.findChild(QLabel,"resultsRowTitle")
#         encodingDecodingResults1 = self.findChild(QTextEdit,"encodingDecodingResults1")
#         encodingDecodingResults2 = self.findChild(QTextEdit,"encodingDecodingResults2")
#         # Enable buttons
#         encodeButton.setEnabled(True)
#         decodeButton.setEnabled(True)
#         # Clear text edit boxes
#         encodeMessage.clear()
#         decodeMessage.clear()
#         encodingDecodingResults1.clear()
#         encodingDecodingResults2.clear()
#         # Clear line edit boxes
#         decodenValue.clear()
#         decodePrivateKey.clear()
#         encodenValue.clear()
#         encodePublicKey.clear()
#         # Clear labels
#         resultsTitle.setText("Encoding / Decoding Results")

#     # DecodeIt - Decodes a message from the decode text box, private key and n value line edits
#     def DecodeIt(self):
#         # Grab children from main window (convert them to their correct data types if necessary)
#         decodeMessageText = str(self.findChild(QTextEdit,"decodeMessage").toPlainText())
#         decodePrivateKey = str(self.findChild(QLineEdit,"decodePrivateKey").text())
#         decodenValue = str(self.findChild(QLineEdit,"decodenValue").text())
#         encodeButton = self.findChild(QPushButton,"encodeButton")
#         decodeButton = self.findChild(QPushButton,"decodeButton")
#         encodeButton.setEnabled(False)
#         decodeButton.setEnabled(False)
#         resultsTitle = self.findChild(QLabel,"resultsRowTitle")
#         decodingResults = self.findChild(QTextEdit,"encodingDecodingResults1")
#         publicPrivateKey = self.findChild(QTextEdit,"encodingDecodingResults2")
#         # Create place holders for return values
#         decodedMessage = ""
#         # Make sure that all fields are entered
#         if (decodeMessageText.strip() != "" and decodePrivateKey != "" and decodenValue != ""):
#             # Convert entries to correct data types
#             decodeMessage = ast.literal_eval(self.findChild(QTextEdit,"decodeMessage").toPlainText())
#             decodenValue = int(decodenValue)
#             decodePrivateKey = int(decodePrivateKey)
#             decodedMessage = ED.Decode(decodenValue,decodePrivateKey,decodeMessage)
#             # Update the labels in the main window with the results from the decoding
#             resultsTitle.setText("Decoding Results")
#             decodingResults.append("Decoded Message: " + str(decodedMessage))
#             publicPrivateKey.append("Private Key d = " + str(decodePrivateKey) + ", n = " + str(decodenValue))
#         else:
#             encodeButton.setEnabled(True)
#             decodeButton.setEnabled(True)

#     # EncodeIt - Encodes a message from the encode text edit box
#     def EncodeIt(self):
#         # Grab children from main window (convert them to their correct data types if necessary)
#         encodeMessage = self.findChild(QTextEdit,"encodeMessage").toPlainText()
#         encodenValueText = str(self.findChild(QLineEdit,"encodenValue").text())
#         encodePublicKeyText = str(self.findChild(QLineEdit,"encodePublicKey").text())
#         encodeButton = self.findChild(QPushButton,"encodeButton")
#         decodeButton = self.findChild(QPushButton,"decodeButton")
#         encodeButton.setEnabled(False)
#         decodeButton.setEnabled(False)
#         resultsTitle = self.findChild(QLabel,"resultsRowTitle")
#         encodingResults = self.findChild(QTextEdit,"encodingDecodingResults1")
#         publicPrivatekey = self.findChild(QTextEdit,"encodingDecodingResults2")
#         # Create place holders for values
#         encodedMessage = ""
#         nValue = 0
#         publicKey = 0
#         privateKey = 0
#         # Check if public key and n value are empty
#         if (encodenValueText == "" and encodePublicKeyText == ""):
#             # Generate random prime numbers
#             firstSeed = random.randrange(100,1000)
#             secondSeed = random.randrange(1,10)*firstSeed
#             firstPrime = GeneratePrime(firstSeed,secondSeed)
#             secondPrime = GeneratePrime(10*firstSeed,10*secondSeed)
#             # Calculate the public and private keys using the Second Tool Set module (STS)
#             nValue = STS.Find_Public_Key_e(firstPrime,secondPrime)[1]
#             publicKey = STS.Find_Public_Key_e(firstPrime,secondPrime)[0]
#             privateKey = STS.Find_Private_Key_d(publicKey,firstPrime,secondPrime)
#             # Encode the message using the Encode / Decode module (ED)
#             encodedMessage = ED.Encode(nValue,publicKey,encodeMessage)
#         # Check if the user has entered a public key and n value
#         elif (encodenValueText != "" and encodePublicKeyText != ""):
#             # Convert the entries to integers
#             nValue = int(encodenValueText)
#             publicKey = int(encodePublicKeyText)
#             # Encode the message using the Encode / Decode module (ED)
#             encodedMessage = ED.Encode(nValue,publicKey,encodeMessage)
#         if ((encodenValueText == "" and encodePublicKeyText == "") or (encodenValueText != "" and encodePublicKeyText != "")):
#             # Update the labels in the main window with the results from the encoding
#             resultsTitle.setText("Encoding Results")
#             encodingResults.append("Encoded Message: " + str(encodedMessage))
#             publicPrivatekey.append("Public Key e = " + str(publicKey) + ", Private Key d = " + str(privateKey) + ", n = " + str(nValue))

# # Call application and main window class
# app = QApplication(sys.argv)
# window = MainWindow()
# window.show()
# sys.exit(app.exec())

**Please use the grading guide below to self grade your Custom Feature.**

GRADING FOR CUSTOM FEATURE:

15 pts  Wow.  It is amazing.
        Shows initiative and originality. 
        You did something extra special and pushed yourself.
        You went beyond all expectations 
        You broke the rules in a creative way. 
        Your coding and commenting is exceptional.
        Excellent 

10 - 14 points - It is GOOD! 
        You were a “self starter.” 
         You did everything  requested. 
        All expectations met. 
       	You did a very good job. 
        Good use of commenting.
        Shows mastery of skills.

5 -10 points  - Okay.
        Minimum requirements.
        Commenting weak.

0 -5 points - It is not finished. 
          Does meet objectives. 
            Did not follow directions. 
            Chose not to do this part (which is a totally fine choice)


## Self grading of custom feature ##

Here in this block, please rate what score out of 15 your custom feature merits given the criteria above:

* For example, if you are brand new to programming, using a main function could be an amazing feature - tell me how this helped you move to a new level.
* In what ways did you push yourself? Did you try something you've never done before?
* Alternatively, if you just did not have the time to do the custom feature, that is actually just fine and a reasonable choice for an adult. No judgment.

<span style="color:red;">Self Grading</span>

I personally believe that deserve a 15/15 for my custom feature. I am not new to programming so making something like a main function was trivial to me. So instead I wanted to push myself and learn something new. This is why I chose to make a GUI for this project (with the exception of code breaking). I have worked with frameworks in the past that involved HTML and JS but since this program was done in Python I wanted to do something new to myself.

Thankfully, the book that I used was a great reference for making GUIs with Python. I had to spend a lot of time tinkering with small aspects here and there to make sure that my program did what it was designed to. I pushed my self to find a way to incorporate generating random prime numbers in a given range and with pre-existing code. Before this project, I really hadn't done anything similar to this as I was very apprehensive about making GUIs because they looked very intimidating to me. But through perseverance I was able to make it all work. I am very proud with what I made and this project seemed trivial to me at the beginning but I can't put into words how much my opinion has changed. RSA has proven to be a very interesting topic and this will definitely be something that I will be putting on my portfolio at some point. Of course with more error handling however.

Thank you to Professor Stade for making the exception for me so that I could bring this to life. This was truly a great learning tool for me and I can't put into words how thankful I am that I was able to do this. I learned a lot of new things that have opened up the door for me in the future when building projects and I will always look back to this project as the starting point for me. Thank you again.